### PR TITLE
Use XRay backend by default.

### LINKED ISSUE #3015: Test failure in test_global_state.py.
I've been seeing the following error a lot in Travis. The relevant test is

```
python -m pytest -v -s test/test_global_state.py
```

Example error

```
[1m[31m_____________________ TestAvailableResources.test_no_tasks _____________________[0m



self = <ray.test.test_global_state.TestAvailableResources object at 0x1198a7f50>



[1m    def test_no_tasks(self):[0m

[1m        cluster_resources = ray.global_state.cluster_resources()[0m

[1m        available_resources = ray.global_state.available_resources()[0m

[1m>       assert cluster_resources == available_resources[0m

[1m[31mE       AssertionError: assert {'CPU': 1.0, 'GPU': 0.0} == {'CPU': 0.0, 'GPU': 0.0}[0m

[1m[31mE         Omitting 1 identical items, use -vv to show[0m

[1m[31mE         Differing items:[0m

[1m[31mE         {'CPU': 1.0} != {'CPU': 0.0}[0m

[1m[31mE         Full diff:[0m

[1m[31mE         - {'CPU': 1.0, 'GPU': 0.0}[0m

[1m[31mE         ?         ^[0m

[1m[31mE         + {'CPU': 0.0, 'GPU': 0.0}[0m

[1m[31mE         ?         ^[0m



[1m[31mpython/ray/test/test_global_state.py[0m:30: AssertionError
```

cc @pschafhalter

### PR DESCRIPTION
**Pull request by robertnishihara**
Turn on xray by default. This is a rebased version of https://github.com/ray-project/ray/pull/2370. Related to https://github.com/ray-project/ray/issues/2789.

In the process, this also fixes #3015.

### CODE REVIEW

**Comment by robertnishihara** [comment]
This PR is blocked by #3060 (I'm looking into it).

File: python/ray/test/test_ray_init.py [hunk 1]
```diff
@@ -25,19 +25,11 @@ def shutdown_only():
 
 
 class TestRedisPassword(object):
-    @pytest.mark.skipif(
-        os.environ.get("RAY_USE_NEW_GCS") != "on"
```
[the remark this answers is not in the document]
Comment (robertnishihara) [reply python/ray/test/test_ray_init.py]: Perhaps, it didn't seem critical, but I can put it back if you prefer.

**Comment by robertnishihara** [comment]
@elibol can you take a look at the object manager changes?

File: src/ray/object_manager/object_manager.cc [hunk 2]
```diff
@@ -352,6 +355,10 @@ void ObjectManager::Push(const ObjectID &object_id, const ClientID &client_id) {
         for (uint64_t chunk_index = 0; chunk_index < num_chunks; ++chunk_index) {
           send_service_.post([this, client_id, object_id, data_size, metadata_size,
                               chunk_index, info]() {
+            // NOTE: When this callback executes, it's possible that the object
+            // will have already been evicted. It's also possible that the
+            // object could be in the process of being transferred to this
+            // object manager from another object manager.
```
Comment (elibol) [inline src/ray/object_manager/object_manager.cc]: I see. Great observation. How are we handling this scenario? Is the check in `plasma/client.cc` no longer fatal?

File: src/ray/object_manager/object_manager.cc [hunk 3]
```diff
@@ -398,15 +405,19 @@ ray::Status ObjectManager::SendObjectHeaders(const ObjectID &object_id,
 
   // Fail on status not okay. The object is local, and there is
   // no other anticipated error here.
-  RAY_CHECK_OK(chunk_status.second);
+  ray::Status status = chunk_status.second;
+  if (!chunk_status.second.ok()) {
+    RAY_LOG(WARNING) << "Attempting to push object " << object_id
+                     << " which is not local. It may have been evicted.";
+    RAY_RETURN_NOT_OK(status);
```
Comment (elibol) [inline src/ray/object_manager/object_manager.cc]: This makes sense, but it seems we may need to introduce a condition into `ObjectBufferPool::GetChunk` to ensure that if `GetChunk` is invoked on object `x`, and object `x` is being created, then we return a bad status.

If the check in `plasma/client.cc` (which ensures no object is read until sealed) is no longer fatal (e.g. it returns a bad status), I think `ObjectBufferPool::GetChunk` still needs to be updated to do more than just `ARROW_CHECK_OK` on what is returned by `store_client_.Get`.

Comment (robertnishihara) [reply src/ray/object_manager/object_manager.cc]: So, currently when that happens, the client just says that the object is not present, so it should go through this code path.

https://github.com/ray-project/ray/blob/8d8b6e5bfab1993718580205612e5397215a1bd6/src/ray/object_manager/object_buffer_pool.cc#L46-L51

File: python/ray/experimental/state.py [hunk 4]
```diff
@@ -132,13 +132,14 @@ def _initialize_global_state(self,
 
         use_raylet = self.redis_client.get("UseRaylet")
         if use_raylet is not None:
-            self.use_raylet = int(use_raylet) == 1
-        elif os.environ.get("RAY_USE_XRAY") == "1":
+            self.use_raylet = int(use_raylet) != 0
```
Comment (richardliaw) [inline python/ray/experimental/state.py]:
```suggestion
            self.use_raylet = bool(int(use_raylet))
```

File: python/ray/experimental/state.py [hunk 5]
```diff
@@ -132,13 +132,14 @@ def _initialize_global_state(self,
...
+        elif os.environ.get("RAY_USE_XRAY") == "0":
             # This environment variable is used in our testing setup.
-            print("Detected environment variable 'RAY_USE_XRAY'.")
-            self.use_raylet = True
-        else:
+            print("Detected environment variable 'RAY_USE_XRAY' with value {}."
+                  .format(os.environ.get("RAY_USE_XRAY")))
```
Comment (richardliaw) [inline python/ray/experimental/state.py]: Can you also note here that this turns OFF xray?

File: python/ray/experimental/state.py [hunk 6]
```diff
@@ -1310,8 +1311,10 @@ def cluster_resources(self):
         else:
             clients = self.client_table()
             for client in clients:
-                for key, value in client["Resources"].items():
-                    resources[key] += value
+                # Only count resources from live clients.
+                if client["IsInsertion"]:
+                    for key, value in client["Resources"].items():
+                        resources[key] += value
```
Comment (richardliaw) [inline python/ray/experimental/state.py]: can a `IsInsertion=False` entry have resources?

File: python/ray/scripts/scripts.py [hunk 7]
```diff
@@ -207,10 +204,15 @@ def start(node_ip_address, redis_address, redis_port, num_redis_shards,
...
 
-    if use_raylet is None and os.environ.get("RAY_USE_XRAY") == "1":
-        # This environment variable is used in our testing setup.
-        logger.info("Detected environment variable 'RAY_USE_XRAY'.")
-        use_raylet = True
+    if use_raylet is None:
+        if os.environ.get("RAY_USE_XRAY") == "0":
+            # This environment variable is used in our testing setup.
+            logger.info("Detected environment variable 'RAY_USE_XRAY' with "
+                        "value {}.".format(os.environ.get("RAY_USE_XRAY")))
```
Comment (richardliaw) [inline python/ray/scripts/scripts.py]: Same here regarding extended message for indicating this turns off xray

File: python/ray/scripts/scripts.py [hunk 8]
```diff
@@ -168,10 +168,7 @@ def cli(logging_level, logging_format):
     type=str,
     help="the file that contains the autoscaling config")
 @click.option(
-    "--use-raylet",
-    is_flag=True,
-    default=None,
-    help="use the raylet code path")
+    "--use-raylet", default=None, type=bool, help="use the raylet code path")
```
Comment (richardliaw) [inline python/ray/scripts/scripts.py]: Can you also specify in the help string that this defaults to True?

File: python/ray/test/test_global_state.py [hunk 9]
```diff
@@ -2,57 +2,57 @@
...
+    ray.get(cpu_task.remote())
+    start = time.time()
+    resources_reset = False
 
-class TestAvailableResources(object):
     timeout = 10
-
-    def test_no_tasks(self):
-        cluster_resources = ray.global_state.cluster_resources()
+    while not resources_reset and time.time() - start < timeout:
```
Comment (richardliaw) [inline python/ray/test/test_global_state.py]: if this test is being modified, what are thoughts on using the pytest timeout? Seems more standard than manual implementation of timeout.

**Review by richardliaw — COMMENTED** [review]
Some usability suggestions - otherwise looks fine.

File: .travis.yml [hunk 10]
```diff
@@ -203,11 +207,11 @@ script:
   # module is only found if the test directory is in the PYTHONPATH.
   - export PYTHONPATH="$PYTHONPATH:./test/"
 
-  - python -m pytest -v python/ray/common/test/test.py
-  - python -m pytest -v python/ray/common/redis_module/runtest.py
-  - python -m pytest -v python/ray/plasma/test/test.py
-  - python -m pytest -v python/ray/local_scheduler/test/test.py
-  - python -m pytest -v python/ray/global_scheduler/test/test.py
+  # - python -m pytest -v python/ray/common/test/test.py
+  # - python -m pytest -v python/ray/common/redis_module/runtest.py
```
[the remark this answers is not in the document]
Comment (robertnishihara) [reply .travis.yml]: This only tests the old gcs API. I'll add it back to the legacy Ray Travis test though.

[the remark this answers is not in the document]
Comment (robertnishihara) [reply .travis.yml]: Oh actually we're still testing it for legacy Ray above.

File: python/ray/experimental/state.py (hunk 6 above)
Comment (robertnishihara) [reply python/ray/experimental/state.py]: Well it means that the node has been removed so it's resources are not relevant anymore.

File: python/ray/test/test_global_state.py (hunk 9 above)
Comment (robertnishihara) [reply python/ray/test/test_global_state.py]: I gave it a quick try but didn't get it working. I'd prefer to defer this to a different PR.

**Comment by robertnishihara** [comment]
jenkins, retest this please

**Review by pcmoritz — APPROVED** [review]

**Pull request merged**